File: src/main/scala/kr/ac/kaist/jsaf/analysis/typing/domain/State.scala
/*******************************************************************************
    Copyright (c) 2012-2013, S-Core, KAIST.
    All rights reserved.

    Use is subject to license terms.

    This distribution may include materials developed by third parties.
 ***************************************************************************** */

package kr.ac.kaist.jsaf.analysis.typing.domain

case class State(heap: Heap, context: Context) {
  /* tuple-like accessor */
  val _1 = heap
  val _2 = context

  /* partial order */
  def <= (that: State) = {
    this.heap <= that.heap && this.context <= that.context
  }

  /* not a partial order */
  def </ (that: State) = {
    !(this.heap <= that.heap) || !(this.context <= that.context)
  }

  /* join */
  def + (that: State) = {
    new State(this.heap + that.heap, this.context + that.context)
  }

  /* meet */
  def <> (that: State) = {
    new State(this.heap <> that.heap, this.context <> that.context)
  }

  def restrict(set: LPSet) = {
    State(this.heap.restrict(set), this.context.restrict(set))
  }

  def restrict(set: LocSet) = {
    State(this.heap.restrict(set), this.context.restrict(set))
  }

  /* for temporal pre-analysis result, make all the properties absentTop. */
  def absentTop() = {
    State(this.heap.absentTop(), this.context)
  }
}

File: src/main/java/kr/ac/kaist/jsaf/analysis/string/automata/State.java
/*******************************************************************************
    Copyright (c) 2013, KAIST.
    All rights reserved.

    Use is subject to license terms.

    This distribution may include materials developed by third parties.
 ******************************************************************************/

package kr.ac.kaist.jsaf.analysis.string.automata;

/**
 * 오토마타의 한 상태를 나타내는 클래스.
 * @author JiminP
 */
public class State {
	private static long gid = 0;
	private long id;
	
	/**
	 * 새 상태를 생성한다.
	 */
	public State(){
		this.id = gid++;
	}
	
	/**
	 * 기존의 존재하는 상태를 복사한다.
	 * @param state
	 */
	public State(State state){
		this.id = state.id;
	}
	
	/**
	 * 상태의 고유한 ID를 반환한다.
	 * @return	상태의 ID값
	 */
	public long getID(){
		return this.id;
	}
	
	@Override public int hashCode(){
		return (int) this.id;
	}
	
	@Override public boolean equals(Object that){
		return (that instanceof State) && this.id == ((State) that).id;
	}
	
	@Override public String toString(){
		return "[State " + this.id + "]";
	}
}
